File: delete-and-earn/delete-and-earn.cpp
class Solution {
public:
    map<int,int>freq;
    int dp[20010][3];
    int deleteAndEarn(vector<int>& nums) {
        int mx = 0;
        for(int i = 0; i < nums.size(); i++) {
            freq[nums[i]]++;
            mx = max(mx, nums[i]);
        }
        
//         for(auto it: freq){
//             printf("%d %d\n", it.first, it.second);
//         }
        
        memset(dp, -1, sizeof dp);
        
        return solve(mx, 0);
    }
    
    int solve(int num, int action) {
        //printf("%d\n", num);
        if(num == 0) {
            return 0;
        }
        
        if(num == 1) {
            return freq[num] * num;
        }
        
        if(dp[num][action] != -1) {
            return dp[num][action];
        }
        
        int q1 = solve(num-1, 0);
        int q2 = (freq[num] * num) + solve(num-2, 1);
        
        return dp[num][action] = max(q1, q2);
    }
};

File: delete-and-earn/delete-and-earn.py
class Solution:
    def __init__(self):
        self.dp = [ [-1 for i in range(3)] for j in range(10010)]
        self.freq = [ 0 for i in range(10010)]
        
    def deleteAndEarn(self, nums: List[int]) -> int:
        mx = 0
        for num in nums:
            self.freq[num] = self.freq[num] + 1
            mx = max(mx, num)
        return self.solve(mx, 0)
    
    def solve(self, num, action):
        if num == 0:
            return 0
        
        if num == 1:
            return self.freq[num] * num
        
        if self.dp[num][action] != -1:
            return self.dp[num][action]
        
        res1 = self.solve(num - 1, 0)
        res2 = (self.freq[num] * num) + self.solve(num - 2, 1)
        
        self.dp[num][action] = max(res1, res2)
        return self.dp[num][action]
            
        